src/local/terrain.rs

```diff
@@ -0,0 +1,58 @@
+use std::mem::MaybeUninit;
+
+use js_sys::Uint8Array;
+
+use crate::{constants::Terrain, objects::RoomTerrain};
+
+use super::{xy_to_terrain_index, RoomXY, ROOM_AREA};
+
+pub struct LocalRoomTerrain {
+    bits: Box<[u8; ROOM_AREA]>,
+}
+
+impl LocalRoomTerrain {
+    /// Gets the terrain at the specified position in this room.
+    pub fn get(&self, xy: RoomXY) -> Terrain {
+        // SAFETY: RoomXY is always a valid coordinate.
+        let byte = unsafe { self.bits.get_unchecked(xy_to_terrain_index(xy)) };
+        // not using Terrain::from_u8() because `0b11` value, wall+swamp, happens
+        // in commonly used server environments (notably the private server default
+        // map), and is special-cased in the engine code; we special-case it here
+        match byte {
+            0b00 => Terrain::Plain,
+            0b01 | 0b11 => Terrain::Wall,
+            0b10 => Terrain::Swamp,
+            _ => panic!("unexpected terrain"),
+        }
+    }
+
+    pub fn new_from_bits(bits: Box<[u8; ROOM_AREA]>) -> Self {
+        Self { bits }
+    }
+}
+
+impl From<RoomTerrain> for LocalRoomTerrain {
+    fn from(terrain: RoomTerrain) -> LocalRoomTerrain {
+        // create an uninitialized array of the correct size
+        let mut data: Box<[MaybeUninit<u8>; ROOM_AREA]> =
+            Box::new([MaybeUninit::uninit(); ROOM_AREA]);
+        // create a Uint8Array mapped to the same point in wasm linear memory as our
+        // uninitialized boxed array
+        let js_buffer =
+            unsafe { Uint8Array::view_mut_raw(data.as_mut_ptr() as *mut u8, ROOM_AREA) };
+        // SAFETY: it's important to not allocate _anything_ in rust memory now that
+        // we've created the Uint8Array pointed at an arbitrary wasm memory location;
+        // we'd overwrite the wrong memory if the array moved due to allocation!
```

Comment:
As discussed in Discord, I believe that this safety requirement is incorrect (only a correctness requirement). However there is likely a safety requirement that the backing data isn't modified while the return value is used, much like `slice::from_raw_parts_mut`.  I have filed [an issue](https://github.com/rustwasm/wasm-bindgen/issues/3612) asking for clarification.

Also the safety comment (whatever its exact requirements and justification are) should be moved above the `unsafe` block.

Reply:
I think the new safety comment should reflect the state of things as we understand it?